### PR TITLE
GFORMS-1803: produce more meaningful specimen links

### PR DESCRIPTION
**Pull request by lxol**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/formtemplate/Verifier.scala [hunk 1]
```diff
@@ -157,6 +158,84 @@ trait Verifier {
...
+        fc.`type` match {
+          case InformationMessage(infoType, infoText) =>
+            fc.copy(`type` = InformationMessage(infoType, updateSpecimenSmartString(infoText)))
+          case otherwise => fc
+        }
+
+      def updateChoices(fc: FormComponent): FormComponent = {
+        import OptionData._
+        fc.`type` match {
+          case c: Choice =>
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/Verifier.scala:180]: There exists `IsChoice` extractor for this

File: app/uk/gov/hmrc/gform/formtemplate/Verifier.scala [hunk 2]
```diff
@@ -157,6 +158,84 @@ trait Verifier {
...
+    def updateSpecimenFormComponent(formComponent: FormComponent): FormComponent = {
+      def updateLabel(fc: FormComponent): FormComponent =
+        fc.copy(label = updateSpecimenSmartString(fc.label))
+
+      def updateHelpText(fc: FormComponent): FormComponent =
+        fc.copy(helpText = fc.helpText.map(updateSpecimenSmartString))
+
+      def updateInfoText(fc: FormComponent): FormComponent =
+        fc.`type` match {
+          case InformationMessage(infoType, infoText) =>
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/Verifier.scala:172]: There exists `IsInformationMessage` extractor for this

File: app/uk/gov/hmrc/gform/formtemplate/Verifier.scala [hunk 3]
```diff
@@ -157,6 +158,84 @@ trait Verifier {
...
+            val updatedOptions = c.options.map {
+              case IndexBased(label, includeIf)        => IndexBased(updateSpecimenSmartString(label), includeIf)
+              case ValueBased(label, value, includeIf) => ValueBased(updateSpecimenSmartString(label), value, includeIf)
+            }
+            val updatedChoiceType = c.copy(options = updatedOptions)
+            fc.copy(`type` = updatedChoiceType)
+          case otherwise => fc
+        }
+      }
+      (updateLabel _ andThen updateHelpText _ andThen updateInfoText _ andThen updateChoices)(formComponent)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/Verifier.scala:190]: Why is this manipulating help text or info text or even a choices?

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala [hunk 4]
```diff
@@ -86,19 +86,25 @@ class FormTemplateService(
...
-          substitutedFormTemplate,
-          booleanExpressionsContext,
-          expressionsContextSubstituted
-        )
-      _                  <- verify(substitutedBooleanExprsFormTemplate)(expressionsContext)
-      formTemplateToSave <- rewrite(substitutedBooleanExprsFormTemplate)
-      _                  <- formTemplateRepo.replace(mkSpecimen(formTemplateToSave))
-      _                  <- formRedirectRepo.deleteByFieldName("redirect", formTemplate._id.value)
+      formTemplateToSave         <- substitute(formTemplate)
+      formTemplateSpecimenToSave <- substitute(specimenSmartStrings(formTemplate))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplateService.scala:105]: This seems to be missing functionality from last point in the description about top level expression being replaced by expression name

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/formtemplate/Verifier.scala (hunk 3 above)
Comment (lxol) [reply app/uk/gov/hmrc/gform/formtemplate/Verifier.scala:190]: the ticket enumerates the places where it should be replaced. I just implemented ACs

**Review by josef-vlach — CHANGES_REQUESTED** [review]
Can we implement this in the spirit of `Substituter`. Ie. let's introduce a type class for manipulation all `SmartString`s on any given `T`.

**Review by josef-vlach — APPROVED** [review]

**Pull request merged**